### DISCUSSION TITLE
Improving C,C++, C# Code Detection

### DISCUSSION THREAD

**Discussion by ValerioOniride** [squidfunk/mkdocs-material#7066] [Q&A]
### Context

After several attempts with the [extra.css](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/#customization) file and experimenting with different language combinations, including C#, I've concluded it's necessary to submit this request

### Description

I would like to request an enhancement for the code block parsing capability, specifically to improve its recognition of key elements within C#, C++, and C code blocks. I suggest augmenting the parser to identify and classify these elements with a similar accuracy and granularity as demonstrated in the example provided in the Vsuals section."

### Related links

- [Requesting a change](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/)
-

### Use Cases

Enhancing the API's code block parsing will improve code clarity for both authors and users, ensuring better communication of intended messages. This change will expedite learning and development within the community, aligning with the project's goal of providing accessible documentation and fostering collaboration.

### Visuals

<img width="572" alt="Screenshot 2024-04-19 113641" src="https://github.com/squidfunk/mkdocs-material/assets/151024241/a070eae3-810e-48a6-8526-0c66c8cada37">

![immagine](https://github.com/squidfunk/mkdocs-material/assets/151024241/e80129e1-8dd0-4087-8903-97deef69c4be)

### Before submitting

- [X] I have read and followed the [change request guidelines](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/).
- [X] I have verified that [my idea is a change request and not a bug report](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#its-not-a-bug-its-a-feature).
- [X] I have ensured that, to the best of my knowledge, [my idea will benefit the entire community](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#benefit-for-the-community).
- [X] I have included relevant links to [the documentation](https://squidfunk.github.io/mkdocs-material/), related [issues](https://github.com/squidfunk/mkdocs-material/issues), and [discussions](https://github.com/squidfunk/mkdocs-material/discussions) to underline the need for my idea.

**Comment by alexvoss** [comment]
Hi, 

the [code highlighting is configurable](https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/?h=highlight#highlight). Which options have you tried and what did you not manage to achieve?

Please note that there is probably nothing to be done in Material for MkDocs per se. Either [Pygments](https://pygments.org/) or [Highlight.js](https://highlightjs.org/) would need to accept a change request. If you decide to contact them, please try to be more specific about what does not work for you and make sure to document what you have tried.

**Comment by ValerioOniride** [comment]
I haven't tried Highlight.js, im using this .yml configuration:
 markdown_extensions:
  - pymdownx.highlight:
      anchor_linenums: true
      use_pygments: true
  - pymdownx.inlinehilite
  - pymdownx.snippets:
      base_path: ["overrides"]
  - pymdownx.blocks.admonition
  - pymdownx.arithmatex:
      generic: true
  - footnotes
  - pymdownx.blocks.details
  - pymdownx.superfences
  - pymdownx.mark
  - pymdownx.keys
  - attr_list
  - pymdownx.emoji:
      emoji_index: !!python/name:material.extensions.emoji.twemoji
      emoji_generator: !!python/name:material.extensions.emoji.to_svg

The issue i'm facing is that i don't like the result of the codeblock highlight, could you help me setting up highlight.js to make a test with?

**Comment by alexvoss** [comment]
All you need is in the documentation but since we don't have an example in our example repository yet, I just put one together. You need to change the code block to the language you want.

[9.5.18-code-highlight-javascript-example.zip](https://github.com/squidfunk/mkdocs-material/files/15041018/9.5.18-code-highlight-javascript-example.zip)

**Comment by squidfunk** [comment]
Converting to discussion, since this is essentially a customization request.

**Comment by kamilkrzyskow** [comment]
Hello @ValerioOniride,
considering that [Unity itself doesn't bother highlighting it properly](https://docs.unity3d.com/ScriptReference/MenuItem.html) on their own docs, I would assume you need to make adjustments yourself. Also the `MenuItem` seems to be a built-in Unity thing, and `ActiveEditorTracker` seems to be your own defined class. Pygments doesn't process the code, it only lexes tokens and none of the above are defined in the base C# syntax. 

Looking at the demo [here](https://pygments.org/demo/?lexer=csharp&style=monokai&formatter=tokens&code=using+UnityEditor%3B%0Ausing+UnityEngine%3B%0Apublic+class+MenuTest+%3A+MonoBehaviour%0A%7B%0A++++%2F%2F+Add+a+menu+item+named+%22Do+Something%22+to+MyMenu+in+the+menu+bar.%0A++++%5BMenuItem%28%22MyMenu%2FDo+Something%22%29%5D%0A++++static+void+DoSomething%28%29%0A++++%7B%0A++++++++Debug.Log%28%22Doing+Something...%22%29%3B%0A++++%7D%0A%0A++++%2F%2F+Add+a+menu+item+named+%22Log+Selected+Transform+Name%22+to+MyMenu+in+the+menu+bar.%0A++++%2F%2F+We+want+this+to+be+validated+menu+item%3A+an+item+that+is+only+enabled+if+specific+conditions+are+met.%0A++++%2F%2F+To+achieve+this%2C+we+use+a+second+function+below+to+validate+the+menu+item.%0A++++%2F%2F+so+it+will+only+be+enabled+if+we+have+a+transform+selected.%0A++++%5BMenuItem%28%22MyMenu%2FLog+Selected+Transform+Name%22%29%5D%0A++++static+void+LogSelectedTransformName%28%29%0A++++%7B%0A++++++++Debug.Log%28%22Selected+Transform+is+on+%22+%2B+Selection.activeTransform.gameObject.name+%2B+%22.%22%29%3B%0A++++%7D%0A%0A++++%2F%2F+Validate+the+menu+item+defined+by+the+function+above.%0A++++%2F%2F+The+%22Log+Selected+Transform+Name%22+menu+item+is+disabled+if+this+function+returns+false.%0A++++%2F%2F+We+tell+the+Editor+that+this+is+a+validation+function+by+decorating+it+with+a+MenuItem+attribute%0A++++%2F%2F+and+passing+true+as+the+second+parameter.%0A++++%2F%2F+This+invokes+the+MenuItem%28string+itemName%2C+bool+isValidateFunction%29+attribute+constructor%0A++++%2F%2F+resulting+in+the+function+being+treated+as+the+validator+for+%22Log+Selected+Transform+Name%22+menu+item.%0A++++%5BMenuItem%28%22MyMenu%2FLog+Selected+Transform+Name%22%2C+true%29%5D%0A++++static+bool+ValidateLogSelectedTransformName%28%29%0A++++%7B%0A++++++++%2F%2F+Return+false+if+no+transform+is+selected.%0A++++++++return+Selection.activeTransform+%21%3D+null%3B%0A++++%7D%0A%0A++++%2F%2F+Add+a+menu+item+named+%22Do+Something+with+a+Shortcut+Key%22+to+MyMenu+in+the+menu+bar%0A++++%2F%2F+and+give+it+a+shortcut+%28ctrl-g+on+Windows%2C+cmd-g+on+macOS%29.%0A++++%5BMenuItem%28%22MyMenu%2FDo+Something+with+a+Shortcut+Key+%25g%22%29%5D%0A++++static+void+DoSomethingWithAShortcutKey%28%29%0A++++%7B%0A++++++++Debug.Log%28%22Doing+something+with+a+Shortcut+Key...%22%29%3B%0A++++%7D%0A%0A++++%2F%2F+Add+a+menu+item+called+%22Double+Mass%22+to+a+Rigidbody%27s+context+menu.%0A++++%5BMenuItem%28%22CONTEXT%2FRigidbody%2FDouble+Mass%22%29%5D%0A++++static+void+DoubleMass%28MenuCommand+command%29%0A++++%7B%0A++++++++Rigidbody+body+%3D+%28Rigidbody%29command.context%3B%0A++++++++body.mass+%3D+body.mass+*+2%3B%0A++++++++Debug.Log%28%22Doubled+Rigidbody%27s+Mass+to+%22+%2B+body.mass+%2B+%22+from+Context+Menu.%22%29%3B%0A++++%7D%0A%0A++++%2F%2F+Add+a+menu+item+to+create+custom+GameObjects.%0A++++%2F%2F+Priority+10+ensures+it+is+grouped+with+the+other+menu+items+of+the+same+kind%0A++++%2F%2F+and+propagated+to+the+hierarchy+dropdown+and+hierarchy+context+menus.%0A++++%5BMenuItem%28%22GameObject%2FMyCategory%2FCustom+Game+Object%22%2C+false%2C+10%29%5D%0A++++static+void+CreateCustomGameObject%28MenuCommand+menuCommand%29%0A++++%7B%0A++++++++%2F%2F+Create+a+custom+game+object%0A++++++++GameObject+go+%3D+new+GameObject%28%22Custom+Game+Object%22%29%3B%0A++++++++%2F%2F+Ensure+it+gets+reparented+if+this+was+a+context+click+%28otherwise+does+nothing%29%0A++++++++GameObjectUtility.SetParentAndAlign%28go%2C+menuCommand.context+as+GameObject%29%3B%0A++++++++%2F%2F+Register+the+creation+in+the+undo+system%0A++++++++Undo.RegisterCreatedObjectUndo%28go%2C+%22Create+%22+%2B+go.name%29%3B%0A++++++++Selection.activeObject+%3D+go%3B%0A++++%7D%0A%7D) the whole `[MenuItem("MyMenu/Log Selected Transform Name")]` is recognized as a `Name.Attribute`, so you could definitely extend the default lexer to change the behaviour to process the insides separately, but note that you'd have to define what keywords are special built-ins to highlight. Similarly, like I did it with the Gothic lexer [here](https://github.com/kamilkrzyskow/gothic-lexer/blob/b5ee7868a7e7397cd129ff1f847daa7d39c0e416/gothic_lexer/daedalus.py#L178-L183), but note that I didn't look at the base C# lexer, so perhaps this isn't the best approach.

I did a quick search for a Pygments Unity Lexer but found nothing, so you could be the first to make one 😃 

I also tried the Highlight.js [demo](https://highlightjs.org/demo), and found that it did a better job with the Unity MenuItem example, properly tokenizing the `[MenuItem("MyMenu/Log Selected Transform Name")]` part.
